File: tasks/implementations/go/tasks/producer_consumer.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "path/filepath"
    "strconv"
)

func RunProducerConsumer(size string, fixturesRoot string) {
    const mod uint64 = 4294967291
    const capacity = 64
    file, _ := os.Open(filepath.Join(fixturesRoot, "generated", "producer_consumer", size+".txt"))
    defer file.Close()
    scanner := bufio.NewScanner(file)
    values := make([]int, 0)
    for scanner.Scan() {
        value, _ := strconv.Atoi(scanner.Text())
        values = append(values, value)
    }
    buffer := make([]int, capacity)
    head, tail, count := 0, 0, 0
    produced := 0
    consumed := uint64(0)
    checksum := uint64(0)
    for produced < len(values) || count > 0 {
        for produced < len(values) && count < capacity {
            buffer[tail] = values[produced]
            tail = (tail + 1) % capacity
            produced++
            count++
        }
        if count > 0 {
            value := buffer[head]
            head = (head + 1) % capacity
            count--
            checksum = (checksum + uint64(value)*(consumed+1)) % mod
            consumed++
        }
    }
    fmt.Println(checksum)
}

File: tasks/implementations/cpp/tasks/producer_consumer.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <cstdint>

void run_producer_consumer(const std::string& size, const std::string& fixturesRoot) {
    constexpr std::uint64_t MOD = 4294967291ULL;
    constexpr int CAPACITY = 64;
    std::ifstream file(fixturesRoot + "/generated/producer_consumer/" + size + ".txt");
    std::vector<int> values;
    int value = 0;
    while (file >> value) values.push_back(value);
    std::vector<int> buffer(CAPACITY, 0);
    int head = 0, tail = 0, count = 0;
    std::size_t produced = 0;
    std::uint64_t consumed = 0;
    std::uint64_t checksum = 0;
    while (produced < values.size() || count > 0) {
        while (produced < values.size() && count < CAPACITY) {
            buffer[tail] = values[produced++];
            tail = (tail + 1) % CAPACITY;
            count++;
        }
        if (count > 0) {
            int current = buffer[head];
            head = (head + 1) % CAPACITY;
            count--;
            checksum = (checksum + static_cast<std::uint64_t>(current) * (consumed + 1ULL)) % MOD;
            consumed++;
        }
    }
    std::cout << checksum << std::endl;
}
